discovery/client.go

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (C) 2024 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package discovery
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/go-did/vc"
+	"github.com/nuts-foundation/nuts-node/audit"
+	nutsCrypto "github.com/nuts-foundation/nuts-node/crypto"
+	"github.com/nuts-foundation/nuts-node/discovery/api/v1/client"
+	"github.com/nuts-foundation/nuts-node/discovery/log"
+	"github.com/nuts-foundation/nuts-node/vcr"
+	"github.com/nuts-foundation/nuts-node/vcr/holder"
+	"github.com/nuts-foundation/nuts-node/vcr/signature/proof"
+	"time"
+)
+
+// registrationManager is responsible for managing registrations on a Discovery Service.
+// It automatically refreshes registered Verifiable Presentations when they are about to expire.
+type registrationManager interface {
+	register(ctx context.Context, serviceID string, subjectDID did.DID) error
+	deregister(ctx context.Context, serviceID string, subjectDID did.DID) error
+	// refreshRegistrations is a blocking call to periodically refresh registrations.
+	// It checks for registrations to be refreshed at the specified interval.
+	// It will exit when the given context is cancelled.
+	refreshVerifiablePresentations(ctx context.Context, interval time.Duration)
+}
+
+var _ registrationManager = &scheduledRegistrationManager{}
+
+type scheduledRegistrationManager struct {
+	services map[string]ServiceDefinition
+	store    *sqlStore
+	client   client.HTTPClient
+	vcr      vcr.VCR
+}
+
+func newRegistrationManager(services map[string]ServiceDefinition, store *sqlStore, client client.HTTPClient, vcr vcr.VCR) *scheduledRegistrationManager {
+	instance := &scheduledRegistrationManager{
+		services: services,
+		store:    store,
+		client:   client,
+		vcr:      vcr,
+	}
+	return instance
+}
+
+func (r *scheduledRegistrationManager) register(ctx context.Context, serviceID string, subjectDID did.DID) error {
```

Comment:
Shall I rename client-side calls (registering a DID for a service, so VPs will be registered on the Discovery Service) to `activate(DID)`?

Reply:
Renamed stuff

Reply:
I agree with the renaming. It does not address this issue though.

Reply:
I swapped the 2 updates around; now it will store `0` first, making the presentation eligible for refreshing at the next interval. That way, if registration fails, it will always be retried as soon as possible (earlier, a db error could break the second call, delaying the retry). Downside is that it causes a functional race condition with `refresh()`, which could lead to 2 presentations being registered for the same DID short after each other. But this won't break anything, so this is the lesser evil.